### DISCUSSION TITLE
progress bar in SD.Next

### DISCUSSION THREAD

**Discussion by minsky91** [vladmandic/sdnext#3149] [Q&A]
I tried SD.Next webui a couple of times under Stability Matrix, it is a very appealing SD platform. One thing about it however puzzled me each time: there appears no progress bar in its interface - the one that would show the entire progress of the generation process, not just that of the current stage (e.g. VAE encoding). Maybe I am missing something?

**Comment by vladmandic** [comment]
and how would you represent that? for example, vae decode may take a long time, but its a single operation and there is no way to see inside of it how much of decode is complete or not.

**Reply by minsky91** [reply]
Well, you could certainly have arranged a basic progress-bar tracking for tile-based processes, like SD Upscale / Ultimate SD Upscale (those that take the longest), couldn’t you? What I see in SD.Next’s UI, it just sits there showing ‘img2img 100% Finishing’ most of the time, until the very end. (Which takes btw a little over 5 minutes in SD.Next to 1 X SD-upscale a 7K image, as compared to just 2.5 minutes in Forge for the same image, checkpoint and all other parameters.) The console log, in the meantime, displays merrily a tile after tile being processed, in a very verbose manner. 

For a non-tiled generation process, the sampling stage takes the longest, and it’s not tracked in SD.Next’s progress bar until 40% of sampling  is done (which can take  quite a long time, with some samplers and parameters), by what I see looking in the console logs. I tried to prepare screenshots to illustrate that with an example of batch-processing, but suddenly SD.Next developed a problem, starting to hang right after completion of the vae decode stage (indicating ‘vae 100% finishing’ in the UI and nothing further) at the end of the batch-generating of 8 1Kx1K images. Restarting SD.Next does not help, it runs into the same problem every and each time. Here is the console log of the above operation, from rebooting till the moment it hangs, in case you want to look into the issue (GPU: nvidia RTX 4070 Ti Super, 16 GB VRAM):

13:45:48-401228 INFO     Starting SD.Next                                      
13:45:48-403236 INFO     Logger: file="C:\AI\StabilityMatrix\Packages\SD.Next  
                         Web UI\sdnext.log" level=INFO size=703003 mode=append 
13:45:48-404823 INFO     Python 3.10.11 on Windows                             
13:45:48-497988 INFO     Version: app=sd.next updated=2024-05-07 hash=e081f232 
                         branch=master                                         
                         url=https://github.com/vladmandic/automatic/tree/maste
                         r                                                     
13:45:48-815386 INFO     Latest published version:                             
                         e2af4542bba594e318b3b2e34e47dbe0c771e3bf              
                         2024-05-21T18:51:30Z                                  
13:45:48-828646 INFO     Platform: arch=AMD64 cpu=Intel64 Family 6 Model 183   
                         Stepping 1, GenuineIntel system=Windows               
                         release=Windows-10-10.0.22631-SP0 python=3.10.11      
13:45:48-830648 INFO     nVidia CUDA toolkit detected: nvidia-smi present      
13:45:48-895109 INFO     Extensions: disabled=['sd-webui-controlnet']          
13:45:48-895109 INFO     Extensions: enabled=['Lora', 'sd-extension-chainner', 
                         'sd-extension-system-info',                           
                         'sd-webui-agent-scheduler',                           
                         'stable-diffusion-webui-images-browser',              
                         'stable-diffusion-webui-rembg'] extensions-builtin    
13:45:48-895109 INFO     Extensions:                                           
                         enabled=['multidiffusion-upscaler-for-automatic1111', 
                         'sd-webui-kohya-hiresfix',                            
                         'ultimate-upscale-for-automatic1111'] extensions      
13:45:48-895109 INFO     Startup: quick launch                                 
13:45:48-895109 INFO     Verifying requirements                                
13:45:48-895109 INFO     Verifying packages                                    
13:45:48-910738 INFO     Extensions: disabled=['sd-webui-controlnet']          
13:45:48-910738 INFO     Extensions: enabled=['Lora', 'sd-extension-chainner', 
                         'sd-extension-system-info',                           
                         'sd-webui-agent-scheduler',                           
                         'stable-diffusion-webui-images-browser',              
                         'stable-diffusion-webui-rembg'] extensions-builtin    
13:45:48-910738 INFO     Extensions:                                           
                         enabled=['multidiffusion-upscaler-for-automatic1111', 
                         'sd-webui-kohya-hiresfix',                            
                         'ultimate-upscale-for-automatic1111'] extensions      
13:45:48-910738 INFO     Command line args: ['--use-cuda'] use_cuda=True       
13:45:52-306851 INFO     Load packages: {'torch': '2.2.0+cu121', 'diffusers':  
                         '0.27.0', 'gradio': '3.43.2'}                         
13:45:53-057816 INFO     Engine: backend=Backend.DIFFUSERS compute=cuda        
                         device=cuda attention="Disabled" mode=no_grad         
13:45:53-107076 INFO     Device: device=NVIDIA GeForce RTX 4070 Ti SUPER n=1   
                         arch=sm_90 cap=(8, 9) cuda=12.1 cudnn=8801            
                         driver=552.22                                         
13:45:53-682843 INFO     Available VAEs: path="models\VAE" items=0             
13:45:53-683846 INFO     Disabled extensions: ['sd-webui-controlnet',          
                         'multidiffusion-upscaler-for-automatic1111']          
13:45:53-684849 INFO     Removing empty folder: models\BSRGAN                  
13:45:53-690845 INFO     Available models: path="models\Stable-diffusion"      
                         items=45 time=0.00                                    
13:45:53-752397 INFO     Extension:                                            
                         script='extensions-builtin\Lora\scripts\lora_script.py
                         ' 13:45:53-752397 INFO       
                         LoRA networks: available=60           
                         folders=2                             
13:45:54-077658 INFO     Extension:                                            
                         script='extensions-builtin\sd-webui-agent-scheduler\sc
                         ripts\task_scheduler.py' Using sqlite file:           
                         extensions-builtin\sd-webui-agent-scheduler\task_sched
                         uler.sqlite3                                          
13:45:54-249146 ERROR    Upscaler unrecognized model: type=RealESRGAN          
                         model=RealESRGAN realesrganX4plusAnime_v1             
13:45:54-249146 ERROR    Upscaler unrecognized model: type=RealESRGAN          
                         model=RealESRGAN realesrganX4plus_v1                  
13:45:54-264149 ERROR    Upscaler unrecognized model: type=RealESRGAN          
                         model=RealESRGAN realesrGeneralX4_v3                  
13:45:54-264149 INFO     UI theme: name="light-teal" style=Auto base=sdnext.css
13:45:55-141403 INFO     Extension list is empty: refresh required             
13:45:55-925254 INFO     Local URL: http://127.0.0.1:7861/                     
13:45:56-185636 INFO     [AgentScheduler] Task queue is empty                  
13:45:56-186623 INFO     [AgentScheduler] Registering APIs                     
13:45:56-257523 INFO     Select: model="juggernautXL_v9Rundiffusionphoto2      
                         [c9e3e68f89]"                                         
13:45:56-289541 INFO     Setting Torch parameters: device=cuda                 
                         dtype=torch.float16 vae=torch.float16                 
                         unet=torch.float16 context=no_grad fp16=True bf16=None
                         optimization=Disabled                                 
13:45:56-289541 INFO     Autodetect: model="Stable Diffusion XL"               
                         class=StableDiffusionXLPipeline                       
                         file="C:\AI\StabilityMatrix\Packages\SD.Next Web      
                         UI\models\Stable-diffusion\juggernautXL_v9Rundiffusion
                         photo2.safetensors" size=6776MB                       
13:45:59-574446 INFO     Load embeddings: loaded=0 skipped=1 time=0.02         
13:45:59-731795 INFO     Load model: time=3.30 load=3.30 native=1024 {'ram':   
                         {'used': 1.14, 'total': 31.85}, 'gpu': {'used': 8.08, 
                         'total': 15.99}, 'retries': 0, 'oom': 0}              
13:45:59-731795 INFO     Startup time: 10.81 torch=2.70 gradio=0.64            
                         libraries=1.34 extensions=0.53 ui-en=0.14             
                         ui-control=0.07 ui-settings=0.15 ui-extensions=1.02   
                         ui-defaults=0.05 launch=0.10 api=0.05 app-started=0.28
                         checkpoint=3.47                                       
13:46:21-076969 INFO     Base: class=StableDiffusionXLPipeline                 
Progress  1.90s/it █████████████████████████████████ 100% 30/30 00:57 00:00 Base

**Comment by vladmandic** [comment]
re: sd upscale - that means that those scripts need to be updated to actually set job states correctly. i can do that for sd upscale, but not ultimate upscale as that is external extension.

re: starts at 40% - no, that is simply a consequence of ui polling intervals, you can set them as high or as low as you want.

re: tiling performance - that's a separate topic. we can look into that, but lets do that separately - combining issues in a single filing never works.
